src/coordinator/block/types.go

```diff
@@ -43,36 +43,101 @@ type SeriesMeta struct {
 	Name string
 }
 
-// Bounds are the time bounds
+// Bounds are the time bounds, start time is inclusive but end is exclusive
 type Bounds struct {
 	Start    time.Time
-	End      time.Time
+	Duration time.Duration
 	StepSize time.Duration
 }
 
 // TimeForIndex returns the start time for a given index assuming a uniform step size
 func (b Bounds) TimeForIndex(idx int) (time.Time, error) {
-	step := b.StepSize
-	t := b.Start.Add(time.Duration(idx) * step)
-	if t.After(b.End) {
+	duration := time.Duration(idx) * b.StepSize
+	if b.Steps() == 0 || duration >= b.Duration {
 		return time.Time{}, fmt.Errorf("out of bounds, %d", idx)
 	}
 
-	return t, nil
+	return b.Start.Add(duration), nil
+}
+
+// End calculates the end time for the block and is exclusive
+func (b Bounds) End() time.Time {
+	return b.Start.Add(b.Duration)
 }
 
 // Steps calculates the number of steps for the bounds
 func (b Bounds) Steps() int {
-	if b.Start.After(b.End) || b.StepSize <= 0 {
+	if b.StepSize <= 0 {
 		return 0
 	}
 
-	return int(b.End.Sub(b.Start)/b.StepSize) + 1
+	return int(b.Duration / b.StepSize)
+}
+
+// Contains returns whether the time lies between the bounds.
+func (b Bounds) Contains(t time.Time) bool {
+	return !b.Start.After(t) && b.Start.Add(b.Duration).After(t)
+}
+
+// Next returns the nth next bound from the current bound
+func (b Bounds) Next(n int ) Bounds {
+	return b.nth(n, true)
+}
+
+// Previous returns the nth previous bound from the current bound
+func (b Bounds) Previous(n int ) Bounds {
+	return b.nth(n, false)
+}
+
+func (b Bounds) nth(n int, forward bool) Bounds {
+	multiplier := 1
+	if !forward {
+		multiplier = -1
+	}
+
+	blockDuration := b.Duration
+	start := b.Start.Add(blockDuration * time.Duration(n*multiplier))
+	return Bounds{
+		Start:    start,
+		Duration: blockDuration,
```

Comment:
can just do `Duration: b.Duration`?

Reply:
b.Duration was being used twice so i just extract it out